### PR TITLE
Allocator improvements

### PR DESCRIPTION
**Pull request by dtmoodie**
## This pullrequest changes

This pullrequest updates the interface to allow setting of a thread local allocator gpu and cpu mats which allows for the setting of a unique allocator instance on each thread.

### CODE REVIEW

**Comment by dtmoodie** [comment]
I'd like to request this improvement merged, since the failing test is unrelated to this PR.

File: modules/core/include/opencv2/core/cvdef.h [hunk 1]
```diff
@@ -298,6 +298,30 @@ enum CpuFeatures {
 #  define CV_VFP 0
 #endif
 
+#ifndef CV_THREAD_LOCAL
```
Comment (alalek) [inline modules/core/include/opencv2/core/cvdef.h]: This is not needed in public headers.

Also we use [cv::TLSData wrapper](https://github.com/opencv/opencv/blob/3.2.0/modules/core/include/opencv2/core/utility.hpp#L634) over of "thread_local".

**Comment by dtmoodie** [comment]
The reason I added this commit was so that I could use specific allocators for sections of an algorithm, but the allocator's lifespan could exist beyond the scope of the thread.  Do you have a suggested approach that would allow this? Or is this just a bad idea in the first place?

**Comment by dtmoodie** [comment]
I've found that using thread local storage for allocators managing GPU memory in a cache does not work.  When my application exists, the cuda runtime shuts down before TLS is cleaned up, in which case the destructor for my allocator throws exceptions when called from TLS cleanup because it is trying to release cuda memory after the cuda runtime has already shutdown.

**Comment by vpisarev** [comment]
ok, let me close it for now

**Pull request closed without merging**